fuzzers/main.py

```diff
@@ -0,0 +1,337 @@
+import datetime
+import google.cloud.storage
+from http.server import BaseHTTPRequestHandler, HTTPServer
+import inotify.adapters
+import os
+import pathlib
+import prometheus_client
+import random
+import signal
+import subprocess
+import sys
+import threading
+import time
+import toml
+import uuid
+
+from lib import config
+from workers import utils
+
+WORKDIR = utils.WORKDIR
+REPO_URL = utils.REPO_URL
+#WORKDIR = pathlib.Path('/tmp')
+#REPO_URL = 'https://github.com/Ekleog/nearcore'
+def connect_to_gcs():
+    subprocess.run(['gcloud', 'auth', 'activate-service-account', '--key-file', GCS_CREDENTIALS_FILE], check=True)
+    return google.cloud.storage.Client.from_service_account_json(GCS_CREDENTIALS_FILE)
+    #return google.cloud.storage.Client(project = 'near-nayduck')
+
+NUM_FUZZERS = os.cpu_count()
+#NUM_FUZZERS = 1
+
+CMD_PORT = 7055
+METRICS_PORT = 5507
+
+REPO_DIR_PARENT = WORKDIR
+REPO_DIR_NAME = 'fuzzed-nearcore'
+REPO_DIR = REPO_DIR_PARENT / REPO_DIR_NAME
+
+GCS_CREDENTIALS_FILE = config.CONFIG_DIR / 'credentials.json'
+GCS_BUCKET = 'fuzzer'
+
+FUZZ_TIME = prometheus_client.Counter('fuzz_seconds', 'Time spent fuzzing', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CRASHES = prometheus_client.Counter('fuzz_crashes', 'Number of times the fuzzer process crashed (not unique crashes)', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_ARTIFACTS_FOUND = prometheus_client.Counter('fuzz_artifacts_found', 'Number of artifacts found (should be number of unique crashes)', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CORPUS_UPLOADED = prometheus_client.Counter('fuzz_corpus_uploaded', 'Number of elements uploaded to GCS corpus', ['branch', 'crate', 'runner', 'flags'])
+FUZZ_CORPUS_DELETED = prometheus_client.Counter('fuzz_corpus_deleted', 'Number of elements deleted from GCS corpus', ['branch', 'crate', 'runner', 'flags'])
+
+def update_repo(branch):
+    if not REPO_DIR.exists():
+        print(f"Doing initial clone of repository {REPO_DIR}")
+        if 0 != subprocess.run(['git', 'clone', REPO_URL, REPO_DIR_NAME], cwd=REPO_DIR_PARENT).returncode:
+            raise RuntimeError(f"Failed to clone from repo {REPO_URL}")
+
+    print(f"Updating to latest commit of branch {branch}")
+    if 0 != subprocess.run(['git', 'checkout', '-f', branch], cwd=REPO_DIR).returncode:
+        raise RuntimeError(f"Failed to checkout branch {branch}")
+    if 0 != subprocess.run(['git', 'pull', '--ff-only', REPO_URL, branch], cwd=REPO_DIR).returncode:
+        raise RuntimeError(f"Failed to fast-forward branch {branch} to latest from repo {REPO_URL}")
```

Comment:
Hmm so the reason why I had done it this way was actually to avoid state in the remote list. I think I've found a way (git fetch + checkout FETCH_HEAD) that relies on neither remote nor branch state, what do you think about it?